src/main/scala/pipedsl/typechecker/LinearExecutionChecker.scala

```diff
@@ -0,0 +1,131 @@
+
+  private val solver: Z3Solver = ctx.mkSolver()
+
+  private val predicates: mutable.Stack[Z3AST] = mutable.Stack(ctx.mkFalse())
+  private var currentPipe: Id = null
+
+  override def emptyEnv(): Environments.Environment[Id, Z3AST] = null
+
+  /* functions do not have any out of order properties, so we don't need to */
+  /* check them */
+  override def checkFunc(f: FuncDef,
+                         env: Environments.Environment[Id, Z3AST])
+  : Environments.Environment[Id, Z3AST] = env
+
+  override def checkModule(m: ModuleDef, env: Environments.Environment[Id,
+    Z3AST]): Environments.Environment[Id, Z3AST] =
+    {
+      predicates.clear()
+      predicates.push(ctx.mkFalse())
+      currentPipe = m.name
+      checkCommand(m.body)
+      checkAllRecurse() match
+      {
+        case Z3Status.SATISFIABLE | Z3Status.UNKNOWN => throw new
+            RuntimeException("There are program paths that do not have " +
+              "output nor call")
+        case _ =>
+      }
+      env
+    }
+
+  override def checkCircuit(c: Circuit, env: Environments.Environment[Id,
+    Z3AST]): Environments.Environment[Id, Z3AST] = env
+
+  def checkCommand(c: Command): Unit =
+    {
+      c match
+      {
+        case CSeq(c1, c2) => checkCommand(c1); checkCommand(c2)
+        case CTBar(c1, c2) => checkCommand(c1); checkCommand(c2)
+        case CIf(_, cons, alt) => checkCommand(cons); checkCommand(alt)
+        case CSplit(cases, default) =>
+          for (caseObj <- cases) checkCommand(caseObj.body)
+          checkCommand(default)
+        case COutput(_) => verifyRecursive(c.predicateCtx.get)
+        case CSpecCall(_, pipe, args) =>
+          val pred = c.predicateCtx.get
+          args.foreach(a => checkExpr(a, pred))
+          if(currentPipe == pipe)
+            verifyRecursive(c.predicateCtx.get)
+        case CAssign(_, rhs) => checkExpr(rhs, c.predicateCtx.get)
+        case CExpr(exp) => checkExpr(exp, c.predicateCtx.get)
+        case _ =>
+      }
+    }
+
+  def checkExpr(e: Expr, predicate: Z3BoolExpr)
+  : Unit = e match
+  {
+    case EUop(_, ex) => checkExpr(ex, predicate)
+    case EBinop(_, e1, e2) => checkExpr(e1, predicate); checkExpr(e2, predicate)
+    case ETernary(cond, tval, fval) =>
+      checkExpr(cond, predicate)
+      checkExpr(tval, predicate)
+      checkExpr(fval, predicate)
+    case EApp(_, args) =>
+      args.foreach(e => checkExpr(e, predicate))
+    case ECast(_, exp) => checkExpr(exp, predicate)
+    case ECall(mod, args) =>
+      args.foreach(a => checkExpr(a, predicate))
+      if(mod == currentPipe)
+        verifyRecursive(predicate)
+    case _ =>
+  }
+
+  /**
+   * Checks to see if [[predicate]] can be true at the same time that ANY of
+   * the globl [[predicates]] are in [[env]] */
+  def checkRecursive(predicate: Z3BoolExpr): Z3Status =
+    {
+      /* we want to know if it is possible to satisfy the current predicate */
+      /* we are testing AND ANY of the other known predicates */
```

Comment:
Making this comment in terms of what you're actually checking
(i.e, predicates represent program paths where there exists a call/output -> so "see if this path is reachable concurrently with any other path")